docs/spikes/2020-04-how-to-migrate-lb3-components.md

```diff
@@ -0,0 +1,626 @@
+   to use introspection and a model setting instead. The trickier part is how to
+   apply changes to models added after the component was mounted.)_
+
+10. For all models with a flag enabled in model settings, setup a custom
+    `afterRemote` hook to modify the HTTP response (e.g. add additional
+    headers).
+
+### REST API
+
+11. Add a new REST API endpoint to every (public) model/entity provided by the
+    app, e.g. `GET /api/{model}/schema` returning JSON schema of the model.
+
+12. Add a new REST API endpoint at a configured path, e.g. `/visualize`
+    returning an HTML page.
+
+13. REST API endpoints providing file upload & download.
+
+14. Add a new local service (e.g. `Ping.ping()`) and expose it via REST API
+    (e.g. `/ping`), allow the user to customize the base path where the API is
+    exposed at (e.g. `/pong`).
+
+15. Intercept REST requests handled by an endpoint contributed by a 3rd-party
+    controller to invoke additional logic before the controller method is called
+    (e.g. automatically create containers when uploading files) and after the
+    controller method returned (e.g. post-process the uploaded file).
+
+16. Add new REST API endpoints using Express `(req, res, next)` style.
+
+### Services (local and remote)
+
+16. Add a new local service (a class providing JS/TS API), e.g. Push service or
+    (database) Migration service.
+
+    - The service must be configurable by the target app (e.g. Apple & Google
+      credentials for push notifications, transport config for logger
+      frameworks, Amazon S3 credentials for storage component, etc.).
+
+      _LB3 typically uses component configuration. I think in LB4, we can use
+      `@configure` specific to each service (see the logger extension for an
+      example)._
+
+17. Inject models (entities & repositories) to the service. These models can be
+    provided either by the extension or by the target application. When using
+    models from the target application, the developer needs an option to specify
+    which models (entities) and associated repositories to use.
+
+18. Provide a service factory to create services dynamically at runtime, e.g.
+    named child loggers created via `app.log('name')` in LB3.
+
+19. Provide a custom service connector requiring (environment-specific)
+    configuration and providing a static service API (static means the API is
+    always the same, as opposed to API generated dynamically from WSDL or
+    OpenAPI).
+
+### API transports
+
+20. A new server type to handle requests coming from a non-REST source, e.g.
+    Primus/WebSockets and RabbitMQ.
+
+21. WebSockets: mount WebSocket-handler on the underlying http server used by
+    `@loopback/rest`.
+
+22. Message queues: allow consumers to also produce (publish) new messages, e.g.
+    as a response to the incoming message.
+
+### Authentication & authorization
+
+21. Add a custom step to the default authentication/authorization sequence, so
```

Comment:
does "the default authentication/authorization sequence" mean LB4 sequence? There is only one action related to authentication currently. The authorization is handled by interceptor.

Reply:
Good question.

By "default authentication/authorization sequence", I mean whatever steps are executed by LB3 runtime to authenticate and authorize an incoming request.

Obviously, this will be different in LB4.

Reply:
I'll change "the default authentication/authorization sequence" to "to the default LB3 authentication/authorization workflow", I hope it will be easier to understand that way.